File: particles/particle_simulator.py
import pygame
import numpy as np
import random


# gravity and electric constanta
G, k = 10000, 1000
fps = 80
type_to_col = {0: (200, 50, 50), 1: (50, 200, 50), 2: (50, 50, 200)}


class Particle:
    def __init__(self, p_type, mass, x, y, vx, vy, charge):
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy
        self.p_type = p_type
        self.mass = mass
        self.charge = charge

    def update_speed(self, dt, other_particles, attraction=[-1, -1], repulsion=[-1, -1]):
        """
        Updates the speed of the particle based on the other particles.
        Takes into consideration gravity, electric forces, and optionally other forces
        :param dt: float
        :param other_particles: list[particle]
            List of all particles
        :param attraction: (int, int)
            defines which different particle types attract each other
        :param repulsion: (int, int)
            defines which different particle types repel each other
        :return: None
        """

        for p in other_particles:
            if p.x == self.x and p.y == self.y:
                continue
            dx = p.x - self.x
            dy = p.y - self.y

            dist = np.sqrt(dx**2 + dy**2)
            r2 = dx**2 + dy**2

            # Electric forces
            # f = q1*q2 / r^2
            # Limiting the maximum force prevents massive accelerations when particles get close
            self.vx += dx * min(k * (p.charge * self.charge) / (r2 * dist), 0.001)
            self.vy += dy * min(k * (p.charge * self.charge) / (r2 * dist), 0.001)

            # other forces
            # Not sure if this part is useful, or if the charges do the same thing
            if p.p_type != self.p_type:
                if p.p_type in attraction and self.p_type in attraction:
                    normalized = (dx / dist, dy / dist)
                    self.vx += normalized[0] * 10 / dist
                    self.vy += normalized[1] * 10 / dist

                if p.p_type in repulsion and self.p_type in repulsion:
                    normalized = (dx / dist, dy / dist)
                    self.vx -= normalized[0] * 10 / dist
                    self.vy -= normalized[1] * 10 / dist

            # Gravitational forces
            # Not allowing too high accelerations is essential, otherwise
            # shit hits fan if particles got too close to each other
            f = min((p.mass * self.mass) * G / r2, 0.2)
            fx = f * dx / (np.abs(dx) + np.abs(dy))
            fy = f * dy / (np.abs(dx) + np.abs(dy))
            self.vx += fx / self.mass * dt
            self.vy += fy / self.mass * dt

    def update_position(self, dt):
        """
        Update the position of the particle based on its
        speed and the time step
        :param dt: float
        :return: None
        """
        self.x += self.vx * dt
        self.y += self.vy * dt
        if 0 > self.x or self.x > 800:
            self.vx = -self.vx
        if 0 > self.y or self.y > 600:
            self.vy = -self.vy


WHITE = (255, 255, 255)
RED = (200, 100, 100)


# Initialize Pygame
pygame.init()

# Set the dimensions of the window
width, height = 800, 600
screen = pygame.display.set_mode((width, height))

# Set a title for the window
pygame.display.set_caption('Pygame Window')
clock = pygame.time.Clock()

particles = []
for i in range(0):
    particles.append(Particle(1, 1, random.randint(0, 800), random.randint(0, 600), random.uniform(-10, 10), random.uniform(-10, 10), 0))

for i in range(0):
    particles.append(Particle(0, 1, random.randint(0, 800), random.randint(0, 600), random.uniform(-10, 10), random.uniform(-10, 10), 0))
particles.append(Particle(0, 1, 400, 300, -8, 0, 0))
particles.append(Particle(1, 1, 400, 400, 8, 0, 0))
particles.append(Particle(0, 1, 300, 300, -8, 0, 0))
particles.append(Particle(1, 1, 500, 400, 8, 0, 0))
particles.append(Particle(0, 1, 250, 350, 0, 5, 0))
particles.append(Particle(1, 1, 550, 350, 0, -5, 0))
particles.append(Particle(0, 1, 350, 350, 0, 7, 0))
particles.append(Particle(1, 1, 450, 350, 0, -7, 0))

running = True
while running:
    screen.fill(WHITE)
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
    # Update the particles
    for p in particles:
        p.update_speed(0.5, particles)
    for p in particles:
        p.update_position(0.5)
    # Draw
    for p in particles:
        pygame.draw.circle(screen, type_to_col[p.p_type], (int(p.x), int(p.y)), 5)
    pygame.display.flip()
    clock.tick(fps)

# Quit Pygame
pygame.quit()

File: particles/particle_simulator.cpp
#include <SFML/Graphics.hpp>
#include <iostream>
#include <cmath>
#include <random>

using namespace std;

// Gravity constant
int G = 200;
float maxForce = 0.3;
float maxSpeed = 10.0;
float friction = 0.002;
int size_x = 900;
int size_y = 900;


class Particle {
public:
    float x, y, vx, vy;
    int p_type;
    float mass;
    float charge;
    sf::CircleShape circle = sf::CircleShape(3);

    Particle(float x, float y, float vx, float vy, int p_type, float mass, float charge) 
    : x(x), y(y), vx(vx), vy(vy), p_type(p_type), mass(mass), charge(charge) {}

    void updatePosition(float dt){
        x += vx * dt;
        y += vy * dt;
        if (50 > x || x > size_x-50) {vx = -vx;}
        if (50 > y || y > size_y-50) {vy = -vy;}
        circle.setPosition(x, y);
    }

    void updateSpeed(float dt, std::vector<Particle>* particles, std::map<std::pair<int, int>, float>* table){
        if (particles != nullptr){
            for (Particle& p : *particles) {
                // Below control the attraction/repellion of the particle
                float attraction = (*table)[{p_type, p.p_type}];
                float dx = p.x - x, dy = p.y - y;
                float tot_vx = 0, tot_vy = 0;
                if (dx != 0 || dy != 0) {
                    float r2 = dx*dx + dy*dy;
                    // too high forces break the simulation, 
                    // in reality particles can't get too close to each other
                    float f = std::min((p.mass * mass) * G / r2, maxForce) * attraction;
                    float fx = f * dx / (abs(dx) + abs(dy));
                    float fy = f * dy / (abs(dx) + abs(dy));
                    tot_vx += (fx / mass * dt);
                    tot_vy += (fy / mass * dt);
                }
                vx = std::min((vx + tot_vx) * (1-friction), maxSpeed);
                vy = std::min((vy + tot_vy) * (1-friction), maxSpeed);

            }
            
        }
    }
};



int generateRandomNumber(int min, int max) {
    std::random_device rd;
    std::mt19937 generator(rd());
    std::uniform_int_distribution<int> distribution(min, max);
    return distribution(generator);
}


float grf(float min, float max) {
    static std::mt19937 engine(std::random_device{}());
    std::uniform_real_distribution<float> distribution(min, max);
    return distribution(engine);
}


int main() {

    // create the particles
    std::vector<Particle> particles = {};
    particles.emplace_back(400, 400, 5, 0, 0, 1, 0);
    particles.emplace_back(600, 400, 0, 5, 0, 1, 0);
    particles.emplace_back(400, 600, 0, -5, 0, 1, 0);
    particles.emplace_back(600, 600, -5, 0, 0, 1, 0);

    
    for(int i = 0; i < 100; i++){
        particles.emplace_back(generateRandomNumber(0, size_x), generateRandomNumber(0, size_y), 0, 0, 0, 1, 0);
    }

    for(int i = 0; i < 100; i++){
        particles.emplace_back(generateRandomNumber(0, size_x), generateRandomNumber(0, size_y), 0, 0, 1, 1, 0);
    }

    for(int i = 0; i < 100; i++){
        particles.emplace_back(generateRandomNumber(0, size_x), generateRandomNumber(0, size_y), 0, 0, 2, 1, 0);
    }

    sf::RenderWindow window(sf::VideoMode(size_x, size_y), "Particles");
    sf::Clock clock;
    sf:: Time accumulator = sf::Time::Zero;
    sf::Time timestep = sf::seconds(1.0f / 60.0f);

    std::map<int, sf::Color> colorDict;
    colorDict[0] = sf::Color::Red;
    colorDict[1] = sf::Color::Green;
    colorDict[2] = sf::Color::Blue;

    std::map<std::pair<int, int>, float> attractionDict;

    float min = -4.0f, max = 4.0f;


    /*
    //SYMMETRIC
    //std::vector<float> rfloats = {grf(min, max), grf(min, max), grf(min, max), grf(min, max), grf(min, max), grf(min, max)};
    //                                1              2, 4           3, 7           5              6,  8           9
    std::vector<float> rfloats = {2, 3, -6, 2, -1.5, 0};
    //                            1  24  37 5  68   9
    std::cout << rfloats[0]<<" "<<rfloats[1]<<" "<<rfloats[2]<< std::endl;
    std::cout << rfloats[1]<<" "<<rfloats[3]<<" "<<rfloats[4]<<std::endl;
    std::cout << rfloats[2]<<" "<<rfloats[4]<<" "<<rfloats[5]<<std::endl;

    attractionDict[{0, 0}] = rfloats[0];
    attractionDict[{0, 1}] = rfloats[1];
    attractionDict[{0, 2}] = rfloats[2];

    attractionDict[{1, 0}] = rfloats[1];
    attractionDict[{1, 1}] = rfloats[3];
    attractionDict[{1, 2}] = rfloats[4];

    attractionDict[{2, 0}] = rfloats[2];
    attractionDict[{2, 1}] = rfloats[4];
    attractionDict[{2, 2}] = rfloats[5];
    */



    std::vector<float> rfloats = {grf(min, max), grf(min, max), grf(min, max), grf(min, max), grf(min, max), grf(min, max), grf(min, max), grf(min, max), grf(min, max)};
    std::cout << rfloats[0]<<" "<<rfloats[1]<<" "<<rfloats[2]<< std::endl;
    std::cout << rfloats[3]<<" "<<rfloats[4]<<" "<<rfloats[5]<<std::endl;
    std::cout << rfloats[6]<<" "<<rfloats[7]<<" "<<rfloats[8]<<std::endl;

    attractionDict[{0, 0}] = rfloats[0];
    attractionDict[{0, 1}] = rfloats[1];
    attractionDict[{0, 2}] = rfloats[2];

    attractionDict[{1, 0}] = rfloats[3];
    attractionDict[{1, 1}] = rfloats[4];
    attractionDict[{1, 2}] = rfloats[5];

    attractionDict[{2, 0}] = rfloats[6];
    attractionDict[{2, 1}] = rfloats[7];
    attractionDict[{2, 2}] = rfloats[8];





    while (window.isOpen()) {
        sf::Event event;
        while (window.pollEvent(event)) {
            if (event.type == sf::Event::Closed)
                window.close();
        }

        sf::Time elapsed = clock.restart();
        accumulator += elapsed;

        // Update at fixed timesteps
        while (accumulator >= timestep) {
            accumulator -= timestep;

            window.clear();
            for (auto& particle : particles) {
                particle.updateSpeed(1, &particles, &attractionDict);
            }
            for (auto& particle : particles) {
                particle.updatePosition(1);
                particle.circle.setFillColor(colorDict[particle.p_type]);
                window.draw(particle.circle);
            }

            window.display();
        }
    }

    return 0;
}
